mysys/my_largepage.c

```diff
@@ -355,158 +310,152 @@ uchar* my_large_malloc_int(size_t *size, myf my_flags)
-              "Warning: Failed to allocate %zu bytes from memory."
-              " errno %d\n", *size, errno);
-    }
   }
-  DBUG_RETURN(ptr);
-}
-#endif /* defined(HAVE_MMAP) && !defined(__linux__) && !defined(_WIN32) */
+#endif
+  /*
+    For ASAN, we need to explicitly unpoison this memory region because the OS
+    may reuse that memory for some TLS or stack variable. It will remain
+    poisoned if it was explicitly poisioned before release. If this happens,
+    we'll have hard to debug false positives like in MDEV-21239.
+    For valgrind, we mark it as UNDEFINED rather than NOACCESS because of the
+    implict reuse possiblility.
+  */
+  else
+    MEM_UNDEFINED(ptr, size);
 
-#ifdef _WIN32
+  DBUG_VOID_RETURN;
+}
 
-/* Windows-specific large pages allocator */
 
-uchar* my_large_malloc_int(size_t *size, myf my_flags)
+/* Linux-specific function to determine the sizes of large pages */
+#ifdef __linux__
+static void my_get_large_page_sizes(size_t sizes[my_large_page_sizes_length])
 {
-  DBUG_ENTER("my_large_malloc_int");
-  void* ptr;
-  DWORD alloc_type= MEM_COMMIT | MEM_RESERVE;
-  size_t orig_size= *size;
+  DIR *dirp;
+  struct dirent *r;
+  int i= 0;
+  DBUG_ENTER("my_get_large_page_sizes");
 
-  if (my_use_large_pages)
+  dirp= opendir("/sys/kernel/mm/hugepages");
+  if (dirp == NULL)
   {
-    alloc_type|= MEM_LARGE_PAGES;
-    /* Align block size to my_large_page_size */
-    *size= MY_ALIGN(*size, (size_t) my_large_page_size);
+    perror("Warning: failed to open /sys/kernel/mm/hugepages");
   }
-  ptr= VirtualAlloc(NULL, *size, alloc_type, PAGE_READWRITE);
-  if (!ptr)
+  else
   {
-    if (my_flags & MY_WME)
+    while (i < my_large_page_sizes_length && (r= readdir(dirp)))
     {
-      fprintf(stderr,
-              "Warning: VirtualAlloc(%zu bytes%s) failed; Windows error %lu\n",
-              *size,
-              my_use_large_pages ? ", MEM_LARGE_PAGES" : "",
-              GetLastError());
+      if (strncmp("hugepages-", r->d_name, 10) == 0)
+      {
+        sizes[i]= strtoull(r->d_name + 10, NULL, 10) * 1024ULL;
+        if (!my_is_2pow(sizes[i]))
+        {
+          fprintf(stderr, "Warning: non-power of 2 large page size (%zu) found,"
+                  " skipping\n", sizes[i]);
+          sizes[i]= 0;
+          continue;
+        }
+        ++i;
+      }
     }
-    *size= orig_size;
-    ptr= VirtualAlloc(NULL, *size, MEM_COMMIT | MEM_RESERVE, PAGE_READWRITE);
-    if (!ptr && my_flags & MY_WME)
+    if (closedir(dirp))
     {
-      fprintf(stderr,
-              "Warning: VirtualAlloc(%zu bytes) failed; Windows error %lu\n",
-              *size, GetLastError());
+      perror("Warning: failed to close /sys/kernel/mm/hugepages");
     }
+    qsort(sizes, i, sizeof(size_t), size_t_cmp);
   }
-
-  DBUG_RETURN(ptr);
+  DBUG_VOID_RETURN;
 }
+#endif
 
-/* Windows-specific large pages deallocator */
 
-my_bool my_large_free_int(void *ptr, size_t size)
+#if defined(HAVE_GETPAGESIZES) && !defined(__linux__)
```

Comment:
Why not `#elif HAVE_GETPAGESIZES`?

I suggest to make it similarly to `my_large_free()`, move before `my_init_large_pages()` and remove all pre-declarations.